src/functions.php

```diff
@@ -459,3 +486,49 @@ function select_server(Manager $manager, array $options): Server
 
     return $manager->selectServer($readPreference);
 }
+
+/**
+ * Performs server selection for an aggregate operation with a write stage. The
+ * $options parameter may be modified by reference if a primary read preference
+ * must be forced due to the existence of pre-5.0 servers in the topology.
+ *
+ * @internal
+ * @see https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#aggregation-pipelines-with-write-stages
+ */
+function select_server_for_aggregate_write_stage(Manager $manager, array &$options): Server
```

Comment:
@kevinAlbs: I changed `all_servers_support_write_stage_on_secondary` to no longer initialize the topology. It now just receives a list of servers directly.

This function is my attempt to incorporate your suggestion from https://github.com/mongodb/mongo-php-library/pull/876#discussion_r761158900.

Reply:
I manually tested this approach with a mixed-version replica set:

 * `localhost:27044` (4.4 primary)
 * `localhost:27050` (5.0 secondary)

To create this replica set, I initialized a two-node replica set with 4.4.x and then took the secondary offline and restarted it with a 5.0.x binary. That was necessary to avoid the error described in [this community forum post](https://www.mongodb.com/community/forums/t/addind-replicaset-members-with-different-versions-4-4-6-and-5-0/118579/2), since a replica set initialized with 5.0.x does not allow adding older servers (due to minWireVersion being bumped).

From there, I ran the following script to insert some data fixtures and attempt execution of an `$out` pipeline on a secondary:

```
<?php

require_once 'vendor/autoload.php';

$client = new MongoDB\Client('mongodb://127.0.0.1:27044/?replicaSet=rs0');

$coll = $client->test->foo;
$coll->drop();
$coll->insertMany([['x' => 1], ['x' => 2], ['x' => 3]], ['writeConcern' => new MongoDB\Driver\WriteConcern(2)]);

$cursor = $coll->aggregate(
    [
        ['$match' => ['x' => ['$gt' => 1]]],
        ['$out' => 'bar'],
    ],
    ['readPreference' => new MongoDB\Driver\ReadPreference('secondary')]
);

echo "\nDocs in test.bar:\n";
foreach ($client->test->bar->find([]) as $doc) {
    printf("%s\n", json_encode($doc));
}
```

I also adding some debugging statements to confirm the server on which `aggregate` is executed. With this patch, the 4.4 primary is always selected. Before this patch, the 5.0 secondary would have been selected.

Next, I modified the read preference to include a non-matching tag set to coerce a server selection failure. Before this patch, a server selection exception would have been thrown. With this patch, the 4.4 primary is still selected as the fall back. This addresses the inadvertent BC break in the original implementation, which I alluded to in [DRIVERS-1969](https://jira.mongodb.org/browse/DRIVERS-1969).